aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+
+        let object_signer = create_signer(id.inner);
+        let guid_creation_num = 0;
+        let transfer_events_guid = guid::create(id.inner, &mut guid_creation_num);
+
+        move_to(
+            &object_signer,
+            Object {
+                guid_creation_num,
+                owner: creator_address,
+                allow_ungated_transfer: true,
+                transfer_events: event::new_event_handle(transfer_events_guid),
+            },
+        );
+        CreatorRef { self: id, can_delete }
+    }
+
+    // Creation helpers
+
+    public fun disallow_ungated_transfer(ref: &CreatorRef) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        object.allow_ungated_transfer = false;
+    }
+
+    public fun generate_delete_ref(ref: &CreatorRef): DeleteRef {
+        assert!(ref.can_delete, error::permission_denied(ECANNOT_DELETE));
+        DeleteRef { self: ref.self }
+    }
+
+    /// Create a signer for the CreatorRef
+    public fun generate_signer(ref: &CreatorRef): signer {
+        create_signer(ref.self.inner)
+    }
+
+    public fun object_id_from_creator_ref(ref: &CreatorRef): ObjectId {
+        ref.self
+    }
+
+    // Signer please...
+
+    /// Create a guid for the object, typically used for events
+    public fun create_guid(object: &signer): guid::GUID acquires Object {
+        let addr = signer::address_of(object);
+        let object_data = borrow_global_mut<Object>(addr);
+        guid::create(addr, &mut object_data.guid_creation_num)
+    }
+
+    /// Generate a new event handle
+    public fun new_event_handle<T: drop + store>(
+        object: &signer,
+    ): event::EventHandle<T> acquires Object {
+        event::new_event_handle(create_guid(object))
+    }
+
+    // Deletion helpers
+
+    public fun delete_ref_id(ref: &DeleteRef): ObjectId {
+        ref.self
+    }
+
+    public fun delete(ref: DeleteRef) acquires Object {
+        let object = move_from<Object>(ref.self.inner);
+        let Object {
+            guid_creation_num: _,
+            owner: _,
+            allow_ungated_transfer: _,
+            transfer_events,
+        } = object;
+        event::destroy_handle(transfer_events);
+    }
+
+    // Extension helpers
+
+    public fun generate_extend_ref(ref: &CreatorRef): ExtendRef {
+        ExtendRef { self: ref.self }
+    }
+
+    /// Create a signer for the ExtendRef
+    public fun generate_signer_for_extending(ref: &ExtendRef): signer {
+        create_signer(ref.self.inner)
+    }
+
+    // Transfer functionality
+
+    public fun generate_transfer_ref(ref: &CreatorRef): TransferRef {
+        TransferRef { self: ref.self }
+    }
+
+    public fun generate_linear_transfer_ref(ref: TransferRef): LinearTransferRef {
+        LinearTransferRef { self: ref.self }
+    }
+
+    public fun transfer_with_ref(ref: LinearTransferRef, to: address) acquires Object {
```

Comment:
I think we should built in a revocation mechanism for transfer. This is otherwise awkward to implement outside, but not that difficult here:

- You generate some token which is embedded in the `TransferRef`.
- `generate_transfer_ref` also returns this token
- You can revoke this token so the next time someone tries to use the TransferRef it aborts
 
This can be simplistically implemented by just maintaining a list of tokens in the Object. Revoking removes it from the list. But perhaps there are much smarter and more efficient cryptographic techniques. @alinush

Reply:
I think I'd benefit from seeing your proposal drawn on a whiteboard.

I also notice that these were probably written before reviewing token.move?

So one model might be...

```
#[resource_group_member(group = common)]
struct Token has key {
 .....
}

#[resource_group_member(group = common)]
struct CreatorTransfer has key {
  transfer_ref: TransferRef,
  creator: address,
}

public fun transwer_as_creator(creator: &signer, to: address, object_id: ObjectId) {
  let transfer_ref = borrow_global_mut<CreatorTransfer>(object_id.inner);
  assert!(transfer_ref.creator == signer::address_of(creator));
  transfer_internal(to, object_id);
}
```

I do think it would be nice to come up with a model where I could compress a lot of TransferRef for distinct objects into a single object, but I'm not as convinced.

The other thing that I'm pondering is -- what is the value of an object if we have to revoke something that cannot be allocated again? For example, imagine I'm the DMV and I give out on-chain licenses. If my account gets hacked:
* I temporarily disable management for a small period of time through some sort of emergency mechanism (that feels kind of awkward -- but I think Max also suggested something similar)
* Nothing happens, the DMV recovers, and it fixes the on-chain data by moving licenses back to their original place
* The DMV says we screwed up too big and starts a new data set and disavows the legacy one.

From the tech side, there's definitely some appeal, though I cannot motivate it easily from the user / product side. It might add more complexity to this. So ideally we have a story around an application.

Reply:
It's rather standard in capabilities and access tokens, and your TransfeRef is really one of those things. Not sure what is unclear here. Revocation is really ubiquitous. We had these use cases in the Diem framework, but did not had time to address them.

One example is that you give out the TransferRef to some third party, lets say some token exchange. In the current model, you basically lost 100% control of the object. Its like you have given it away already. Revoke would allow you to cancel such a delegation without needing to involve the other party. Indeed whether a transfer ref is revocable needs to be part of the contract and/or checkable. (`is_revoked(&tref)`). Here is a code example.

```
struct State {
   id: ObjectId,
   revoke_token: object::TransferRevocation
}

public fun request_transfer(s: &signer, id: ObjectId): TransferRef {
  let owner_ref = ...;
  let (ref, revoke_token) = object::generate_transfer_token(&owner_ref);
  move_to<State>(s, State{id, revoke_token}); 
  r
}

public entry revoke(s: &signer) {
  let State{id, revoke_token} = move_from<State>(address_of(s));
  object::revoke(id, revoke_token); 
}
```

Reply:
The challenge with this model is that you end up placing Transfer into a resource and complicate the programmability.

Reply:
Note, that one can build a revocable model on top of this. The creator dictates the behavior anyway and can decide that they want it to be revocable, they design it in that fashion.  So if they want revocable, they place the TransferRef into a resource (key) container where they define the module that allows the owner to use as they please, while the creator can dictate the revocation policy.

Reply:
Generally speaking if a feature seemed like there were many implementations, I either tried to generalize toward something that could provide the basis for all of them or ensured that this object model wouldn't prevent a developer from adding them later. That way if the community drove towards a common practice, we could adopt that as a framework standard.

So let's poke on this one a bit, since it feels like there's no generalizable solution.

What or who are we protecting our users from? TransferRef and hence revocation should defined by the creator of the object. Certainly the creator can leave management of these to the user, but this requires that the creator build a model for the TransferRef to be used in this fashion. I agree conceptually revocation is crucial, but only if we can define revocation from whom, because the revocation is itself a capability and we have to define who can perform that. We then have to construct the ability to be revokable.

I don't see a one size fits all solution here. I'd be fine adding an example of a revokable transfer ref for a simplistic model. I don't see a generalizable revokable model and implementing one solution in the core framework could confuse other efforts or make auditors and developers confused.

For example, the diem revocation model depended on ownership of a specific Struct type: https://github.com/diem/diem/blob/main/diem-move/diem-framework/move-stdlib/sources/Capability.move#L167

We could also make it based upon
* ownership of a specific struct instance
* the address of the signer submitting it
* an ACL list in the object itself

None of these are really generalizable and I think an application specific revocation model makes the most sense. But taking the "ownership of a specific struct instance", this would require placing the revocation ref somewhere and giving access to someone. It also requires placing the TransferRef into a resource within the object module and another capability struct to access it. You cannot revoke the indirect capability but you can revoke the direct capability. Of course if this copyable and there aren't events, it becomes pretty easy to evade ever really being able to revoke.

Reply:
I don't understand why we have such a complex discussion. Revocation is a rather straightforward concept. I sketched a canonical implementation above. If you don't use it, there should no overhead. Revocation is part of the core of many capability based access control systems. It is also part of OAuth. It is build into the core because its part of the fundamental semantics of authorization tokens. And thats how we should few it. Not as an add-on, but a fundamental concept.

Reply:
I'm happy to add an approach on the side, even in the framework. However that feels like an anti-pattern since it doesn't generalize, it's just an example or one such method. My point is that the outlined approach indicates a specific usage. It's incompatible with many approaches, such as leaving a transfer ref in an attached resource for the creator to manage assets.

Let me sketch up some code as an example. From there we can decide where it lands.